lib/libmbedtls/core/cmac.c

```diff
@@ -0,0 +1,253 @@
+ */
+
+#include <assert.h>
+#include <crypto/crypto.h>
+#include <crypto/crypto_impl.h>
+#include <kernel/panic.h>
+#include <mbedtls/cipher.h>
+#include <mbedtls/cmac.h>
+#include <stdlib.h>
+#include <string.h>
+#include <tee_api_types.h>
+#include <utee_defines.h>
+#include <util.h>
+
+struct mbed_cmac_ctx {
+	struct crypto_mac_ctx mac_ctx;
+	mbedtls_cipher_context_t cipher_ctx;
+};
+
+static const struct crypto_mac_ops mbed_des3_cmac_ops;
+static const struct crypto_mac_ops mbed_aes_cmac_ops;
+
+static struct mbed_cmac_ctx *to_cmac_ctx(struct crypto_mac_ctx *ctx, const mbedtls_cipher_id_t cipher_id)
+{
+	assert(ctx);
+
+	if (cipher_id == MBEDTLS_CIPHER_ID_AES)
+		assert(ctx->ops == &mbed_aes_cmac_ops);
+
+	if (cipher_id == MBEDTLS_CIPHER_ID_3DES)
+		assert(ctx->ops == &mbed_des3_cmac_ops);
+
+	return container_of(ctx, struct mbed_cmac_ctx, mac_ctx);
+}
+
+static TEE_Result mbed_cmac_init(struct crypto_mac_ctx *ctx,
+		const uint8_t *key, size_t len, const mbedtls_cipher_id_t cipher_id)
+{
+	if (cipher_id != MBEDTLS_CIPHER_ID_3DES && cipher_id != MBEDTLS_CIPHER_ID_AES)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	struct mbed_cmac_ctx *c = to_cmac_ctx(ctx, cipher_id);
+
+	const mbedtls_cipher_info_t *cipher_info = NULL;
+
+	cipher_info = mbedtls_cipher_info_from_values(cipher_id,
+			len * 8,
+			MBEDTLS_MODE_ECB);
+	if (!cipher_info)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	if (mbedtls_cipher_setup_info(&c->cipher_ctx, cipher_info))
+		return TEE_ERROR_BAD_STATE;
+
+	if (mbedtls_cipher_cmac_reset(&c->cipher_ctx))
+		return TEE_ERROR_BAD_STATE;
+
+	if (mbedtls_cipher_cmac_starts(&c->cipher_ctx, key, len * 8))
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result mbed_des3_cmac_init(struct crypto_mac_ctx *ctx,
+		const uint8_t *key, size_t len)
+{
+	return mbed_cmac_init(ctx, key, len, MBEDTLS_CIPHER_ID_3DES);
+}
+
+static TEE_Result mbed_aes_cmac_init(struct crypto_mac_ctx *ctx,
+		const uint8_t *key, size_t len)
+{
+	return mbed_cmac_init(ctx, key, len, MBEDTLS_CIPHER_ID_AES);
+}
+
+static TEE_Result mbed_des3_cmac_update(struct crypto_mac_ctx *ctx,
+		const uint8_t *data, size_t len)
+{
+	if (mbedtls_cipher_cmac_update(&to_cmac_ctx(ctx, MBEDTLS_CIPHER_ID_3DES)->cipher_ctx,
+				data, len))
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result mbed_aes_cmac_update(struct crypto_mac_ctx *ctx,
+		const uint8_t *data, size_t len)
+{
+	if (mbedtls_cipher_cmac_update(&to_cmac_ctx(ctx, MBEDTLS_CIPHER_ID_AES)->cipher_ctx,
+				data, len))
+		return TEE_ERROR_BAD_STATE;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result mbed_cmac_final(struct crypto_mac_ctx *ctx, uint8_t *digest,
+		size_t len, const mbedtls_cipher_id_t cipher_id)
+{
+	struct mbed_cmac_ctx *c = to_cmac_ctx(ctx, cipher_id);
+
+	uint8_t block_size = cipher_id == MBEDTLS_CIPHER_ID_AES ? TEE_AES_BLOCK_SIZE : TEE_DES_BLOCK_SIZE;
+	uint8_t block_digest[TEE_AES_BLOCK_SIZE] = { };
+
+	uint8_t *tmp_digest = NULL;
+
```

Comment:
Add:
```C
	COMPILE_TIME_ASSERT(TEE_AES_BLOCK_SIZE >= TEE_DES_BLOCK_SIZE);

```
Although it is quite unlikely that block size will ever change :wink: it clearly documents that the declaration of `block_digest` is correct.